File: node-package/src/utils/logger.ts
/**
 * Vibe CLI Logger Module
 * 
 * A consistent logging interface for the Vibe CLI application.
 */
// Use require instead of import to avoid TypeScript issues
const fs = require('fs');
const path = require('path');
const os = require('os');
const process = require('process');

// Define log levels
export enum LogLevel {
  TRACE = 5,
  DEBUG = 10,
  INFO = 20,
  WARNING = 30,
  ERROR = 40,
  CRITICAL = 50
}

// ANSI color codes for terminal output
const COLORS = {
  reset: '\x1b[0m',
  black: '\x1b[30m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  bold: '\x1b[1m'
};

// Define log level colors
const LEVEL_COLORS = {
  [LogLevel.CRITICAL]: COLORS.red + COLORS.bold,
  [LogLevel.ERROR]: COLORS.red,
  [LogLevel.WARNING]: COLORS.yellow,
  [LogLevel.INFO]: COLORS.green,
  [LogLevel.DEBUG]: COLORS.blue,
  [LogLevel.TRACE]: COLORS.magenta
};

// Log level names
const LEVEL_NAMES = {
  [LogLevel.CRITICAL]: 'CRITICAL',
  [LogLevel.ERROR]: 'ERROR',
  [LogLevel.WARNING]: 'WARNING',
  [LogLevel.INFO]: 'INFO',
  [LogLevel.DEBUG]: 'DEBUG',
  [LogLevel.TRACE]: 'TRACE'
};

interface LogOptions {
  /** Enable/disable colorized output for terminal */
  useColors: boolean;
  /** Enable/disable file logging */
  logToFile: boolean;
  /** Directory to store log files */
  logDir?: string;
  /** Minimum log level to display */
  logLevel: LogLevel;
}

/**
 * Main logger class for Vibe CLI
 */
export class Logger {
  private static instance: Logger;
  private loggers: Map<string, Logger> = new Map();
  private options: LogOptions;
  private logFile?: string;
  private name: string;
  private fileStream?: any; // File stream for logging
  
  /**
   * Create a new logger instance
   * 
   * @param name Logger name (typically module name)
   * @param options Logging options
   */
  private constructor(name: string, options?: Partial<LogOptions>) {
    this.name = name;
    
    // Default options
    this.options = {
      useColors: process.stdout.isTTY,
      logToFile: process.env.VIBE_LOG_TO_FILE === 'true',
      logLevel: this.getLogLevelFromEnv(),
      ...options
    };
    
    // Set up file logging if enabled
    if (this.options.logToFile) {
      this.setupFileLogging();
    }
  }
  
  /**
   * Get log level from environment variable
   */
  private getLogLevelFromEnv(): LogLevel {
    const envLevel = process.env.VIBE_LOG_LEVEL?.toUpperCase();
    switch (envLevel) {
      case 'TRACE': return LogLevel.TRACE;
      case 'DEBUG': return LogLevel.DEBUG;
      case 'INFO': return LogLevel.INFO;
      case 'WARNING': return LogLevel.WARNING;
      case 'ERROR': return LogLevel.ERROR;
      case 'CRITICAL': return LogLevel.CRITICAL;
      default: return LogLevel.INFO;
    }
  }
  
  /**
   * Set up file logging
   */
  private setupFileLogging(): void {
    try {
      // Determine log directory
      const homeDir = os.homedir();
      const logDir = this.options.logDir || path.join(homeDir, '.vibe-cloud', 'logs');
      
      // Create log directory if it doesn't exist
      if (!fs.existsSync(logDir)) {
        fs.mkdirSync(logDir, { recursive: true });
      }
      
      // Create log file with timestamp
      const timestamp = new Date().toISOString().replace(/[:.]/g, '').replace('T', '_').slice(0, 15);
      this.logFile = path.join(logDir, `vibe_cli_${timestamp}.log`);
      
      // Create file stream
      this.fileStream = fs.createWriteStream(this.logFile, { flags: 'a' });
    } catch (error) {
      console.error(`Failed to set up file logging: ${(error as Error).message}`);
    }
  }
  
  /**
   * Get a singleton instance of the logger
   */
  public static getInstance(name: string = 'vibe'): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger(name);
    }
    return Logger.instance;
  }
  
  /**
   * Get a logger with a specific name
   */
  public getLogger(name: string): Logger {
    if (!this.loggers.has(name)) {
      const logger = new Logger(name, {
        useColors: this.options.useColors,
        logToFile: this.options.logToFile,
        logDir: this.options.logDir,
        logLevel: this.options.logLevel
      });
      this.loggers.set(name, logger);
    }
    return this.loggers.get(name)!;
  }
  
  /**
   * Format a log message
   */
  private formatLogMessage(level: LogLevel, message: string): string {
    const timestamp = new Date().toISOString();
    const levelName = LEVEL_NAMES[level];
    
    if (this.options.useColors) {
      const color = LEVEL_COLORS[level];
      return `${timestamp} - ${color}${levelName}${COLORS.reset} - ${this.name} - ${message}`;
    } else {
      return `${timestamp} - ${levelName} - ${this.name} - ${message}`;
    }
  }
  
  /**
   * Write a log message
   */
  private log(level: LogLevel, message: string, ...args: any[]): void {
    // Check if this log level should be displayed
    if (level < this.options.logLevel) {
      return;
    }
    
    // Format message with arguments
    let formattedMessage = message;
    if (args.length > 0) {
      // Simple printf-style formatting (replace %s, %d, etc.)
      let i = 0;
      formattedMessage = message.replace(/%[sdj%]/g, (match) => {
        if (match === '%%') return '%';
        return String(args[i++]);
      });
    }
    
    const logMessage = this.formatLogMessage(level, formattedMessage);
    
    // Log to console
    switch (level) {
      case LogLevel.CRITICAL:
      case LogLevel.ERROR:
        console.error(logMessage);
        break;
      case LogLevel.WARNING:
        console.warn(logMessage);
        break;
      case LogLevel.INFO:
        console.info(logMessage);
        break;
      case LogLevel.DEBUG:
      case LogLevel.TRACE:
        console.debug(logMessage);
        break;
    }
    
    // Log to file if enabled
    if (this.fileStream) {
      // Remove color codes for file logging
      const plainMessage = logMessage.replace(/\x1b\[\d+m/g, '');
      this.fileStream.write(plainMessage + '\n');
    }
  }
  
  /**
   * Set the log level
   */
  public setLevel(level: LogLevel): void {
    this.options.logLevel = level;
  }
  
  /**
   * Log methods for different levels
   */
  public trace(message: string, ...args: any[]): void {
    this.log(LogLevel.TRACE, message, ...args);
  }
  
  public debug(message: string, ...args: any[]): void {
    this.log(LogLevel.DEBUG, message, ...args);
  }
  
  public info(message: string, ...args: any[]): void {
    this.log(LogLevel.INFO, message, ...args);
  }
  
  public warning(message: string, ...args: any[]): void {
    this.log(LogLevel.WARNING, message, ...args);
  }
  
  public error(message: string, ...args: any[]): void {
    this.log(LogLevel.ERROR, message, ...args);
  }
  
  public critical(message: string, ...args: any[]): void {
    this.log(LogLevel.CRITICAL, message, ...args);
  }
}

// Create default logger instance
const defaultLogger = Logger.getInstance();

// Export convenience functions
export function getLogger(name?: string): Logger {
  return name ? defaultLogger.getLogger(name) : defaultLogger;
}

export const trace = (message: string, ...args: any[]): void => defaultLogger.trace(message, ...args);
export const debug = (message: string, ...args: any[]): void => defaultLogger.debug(message, ...args);
export const info = (message: string, ...args: any[]): void => defaultLogger.info(message, ...args);
export const warning = (message: string, ...args: any[]): void => defaultLogger.warning(message, ...args);
export const error = (message: string, ...args: any[]): void => defaultLogger.error(message, ...args);
export const critical = (message: string, ...args: any[]): void => defaultLogger.critical(message, ...args);

// Export default logger
export default defaultLogger;

File: python-package/vibe/utils/logger.py
"""
Vibe CLI Logger Module

A consistent logging interface for the Vibe CLI application.
"""
import os
import sys
import logging
from datetime import datetime
from typing import Optional, Dict, Any
from enum import IntEnum

# Define LogLevel enum for type safety
class LogLevel(IntEnum):
    CRITICAL = 50
    ERROR = 40
    WARNING = 30
    INFO = 20
    DEBUG = 10
    TRACE = 5

# Log levels
CRITICAL = logging.CRITICAL  # 50
ERROR = logging.ERROR        # 40
WARNING = logging.WARNING    # 30
INFO = logging.INFO          # 20
DEBUG = logging.DEBUG        # 10
TRACE = 5                    # Custom level below DEBUG

# Configure custom TRACE level
logging.addLevelName(TRACE, "TRACE")

# ANSI color codes for terminal output
COLORS = {
    'reset': '\033[0m',
    'black': '\033[30m',
    'red': '\033[31m',
    'green': '\033[32m',
    'yellow': '\033[33m',
    'blue': '\033[34m',
    'magenta': '\033[35m',
    'cyan': '\033[36m',
    'white': '\033[37m',
    'bold': '\033[1m'
}

# Define log level colors
LEVEL_COLORS = {
    CRITICAL: COLORS['red'] + COLORS['bold'],
    ERROR: COLORS['red'],
    WARNING: COLORS['yellow'],
    INFO: COLORS['green'],
    DEBUG: COLORS['blue'],
    TRACE: COLORS['magenta']
}

class ColoredFormatter(logging.Formatter):
    """Custom formatter to add colors to log messages in terminal output"""
    
    def format(self, record):
        levelno = record.levelno
        levelname = record.levelname
        
        # Add color to level name if terminal supports it
        if sys.stdout.isatty():  # Only use colors for terminal output
            color = LEVEL_COLORS.get(levelno, COLORS['reset'])
            record.levelname = f"{color}{levelname}{COLORS['reset']}"
        
        return super().format(record)

class VibeLogger:
    """Vibe CLI Logger class"""
    
    _instance = None
    _loggers: Dict[str, logging.Logger] = {}
    
    def __new__(cls):
        """Singleton pattern to ensure only one logger instance exists"""
        if cls._instance is None:
            cls._instance = super(VibeLogger, cls).__new__(cls)
            cls._instance._setup_logger()
        return cls._instance
    
    def _setup_logger(self):
        """Set up the logging configuration"""
        # Get log level from environment or use INFO as default
        log_level_name = os.environ.get('VIBE_LOG_LEVEL', 'INFO').upper()
        log_level = getattr(logging, log_level_name, INFO)
        
        # Create console handler
        console_handler = logging.StreamHandler()
        
        # Create formatter with timestamp, level, and message
        console_format = '%(asctime)s - %(levelname)s - %(name)s - %(message)s'
        date_format = '%Y-%m-%d %H:%M:%S'
        formatter = ColoredFormatter(console_format, date_format)
        console_handler.setFormatter(formatter)
        
        # Configure root logger
        root_logger = logging.getLogger('vibe')
        root_logger.setLevel(log_level)
        
        # Clear any existing handlers to avoid duplicate logs
        if root_logger.handlers:
            root_logger.handlers.clear()
        
        root_logger.addHandler(console_handler)
        
        # Create log directory for file logging if enabled
        if os.environ.get('VIBE_LOG_TO_FILE', 'false').lower() == 'true':
            self._setup_file_logging(root_logger, log_level)
    
    def _setup_file_logging(self, root_logger, log_level):
        """Set up file logging if enabled"""
        # Determine log directory
        home_dir = os.path.expanduser('~')
        log_dir = os.path.join(home_dir, '.vibe-cloud', 'logs')
        
        # Create log directory if it doesn't exist
        os.makedirs(log_dir, exist_ok=True)
        
        # Create log file with timestamp
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        log_file = os.path.join(log_dir, f'vibe_cli_{timestamp}.log')
        
        # Create file handler
        file_handler = logging.FileHandler(log_file)
        file_handler.setLevel(log_level)
        
        # Create formatter for file logs
        file_format = '%(asctime)s - %(levelname)s - %(name)s - %(message)s'
        file_formatter = logging.Formatter(file_format)
        file_handler.setFormatter(file_formatter)
        
        # Add file handler to root logger
        root_logger.addHandler(file_handler)
    
    def get_logger(self, name: str = 'vibe') -> logging.Logger:
        """Get a logger instance with the specified name"""
        if name not in self._loggers:
            logger = logging.getLogger(name)
            
            # Add trace method to logger
            def trace(msg, *args, **kwargs):
                if logger.isEnabledFor(TRACE):
                    logger._log(TRACE, msg, args, **kwargs)
            logger.trace = trace
            
            self._loggers[name] = logger
        
        return self._loggers[name]

# Convenience functions to get a logger and log messages

def get_logger(name: Optional[str] = None) -> logging.Logger:
    """Get a logger instance with the specified name"""
    logger_name = name if name else 'vibe'
    return VibeLogger().get_logger(logger_name)

# Global logger instance for direct imports
logger = get_logger()

# Export convenience functions that match the global log levels
def critical(msg, *args, **kwargs):
    """Log a critical message"""
    logger.critical(msg, *args, **kwargs)

def error(msg, *args, **kwargs):
    """Log an error message"""
    logger.error(msg, *args, **kwargs)

def warning(msg, *args, **kwargs):
    """Log a warning message"""
    logger.warning(msg, *args, **kwargs)

def info(msg, *args, **kwargs):
    """Log an info message"""
    logger.info(msg, *args, **kwargs)

def debug(msg, *args, **kwargs):
    """Log a debug message"""
    logger.debug(msg, *args, **kwargs)

def trace(msg, *args, **kwargs):
    """Log a trace message (more detailed than debug)"""
    logger.trace(msg, *args, **kwargs)

# Set logging level from environment variable
def set_level(level):
    """Set the logging level for the root logger"""
    root_logger = logging.getLogger('vibe')
    root_logger.setLevel(level)
    
    # Update all handlers
    for handler in root_logger.handlers:
        handler.setLevel(level)
